File: case-studies/birthdays/scripts/reloo.jl
dirname(Base.active_project()) != pwd() && exit(run(`$(pwd())/julia.sh $(@__FILE__)`).exitcode)
include(joinpath(pwd(), "scripts/utils.jl"))

function handle(input, output; idx, kwargs...)
    rv = JSON.parsefile(input[1])
    rv["stan"]
    rv["stan"]["n_observations"] -= 1
    idx = parse(Int64, idx)
    for key in ["x", "y", "day_of_week", "day_of_year"]
        rv["stan"][key] = vcat(rv["stan"][key][1:idx-1], rv["stan"][key][idx+1:end])
    end
    rv
end

auto_snakemake(snakemake, handle) 

File: case-studies/birthdays/scripts/reloo.py
from utils import *
import cmdstanpy, time


def handle(input, output, cfg, **kwargs):
    model = cmdstanpy.CmdStanModel(exe_file=input[0])
    data = load_json(input[1])
    qoi = load_json(input[2])
    pk = np.array(qoi["pareto_k"])
    idxs = np.flatnonzero(pk > .7)
    if len(idxs) == 0 or len(idxs) > 2 or cfg.find("posthoc") == -1:
        return qoi
    for idx in idxs:
        data["stan"]["loo"] = 1+idx
        fit = model.sample(
            no_string_values(data["stan"]), seed=0, show_progress=True, chains=4, refresh=1, **data["sample"]
        )
        qoi["pareto_k"][idx] = 0
        qoi["cv_elpd"][idx] = np.mean(fit.loo_likelihood)
    return qoi

    # df = fit.draws_pd()
    # draws = df.filter(regex=("_raw")).to_numpy()
    # lp = df.lp__.to_numpy()
    # n_steps = df.n_leapfrog__.to_numpy()
    # divergent = df.divergent__.to_numpy()

    # return dict(
    #     draws=draws,
    #     lp=lp,
    #     n_steps=n_steps,
    #     divergent=divergent,
    # )
    
    
if __name__ == "__main__": auto_snakemake(snakemake, handle)
 